group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -3022,143 +3242,2087 @@ public void testOnLoaded() {
-        assignment.forEach(topicPartitions -> {
-            assignmentMap.put(topicPartitions.topicId(), new HashSet<>(topicPartitions.partitions()));
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withDefaultProtocolTypeAndProtocols()
+            .withReason("exceed max group size")
+            .build();
+
+        IntStream.range(0, 10).forEach(i -> {
+            CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+            CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, responseFuture);
+            assertFalse(responseFuture.isDone());
+            assertTrue(result.records().isEmpty());
         });
-        return assignmentMap;
+
+        CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+        CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, responseFuture);
+        assertTrue(result.records().isEmpty());
+        assertTrue(responseFuture.isDone());
+        assertEquals(Errors.GROUP_MAX_SIZE_REACHED.code(), responseFuture.get(5, TimeUnit.SECONDS).errorCode());
     }
 
-    private void assertRecordsEquals(
-        List<Record> expectedRecords,
-        List<Record> actualRecords
-    ) {
-        try {
-            assertEquals(expectedRecords.size(), actualRecords.size());
+    @Test
+    public void testDynamicMembersJoinGroupWithMaxSizeAndRequiredKnownMember() {
+        boolean requiredKnownMemberId = true;
+        int groupMaxSize = 10;
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .withGenericGroupMaxSize(groupMaxSize)
+            .withGenericGroupInitialRebalanceDelayMs(50)
+            .build();
+
+        GenericGroup group = context.createGenericGroup("group-id");
+
+        JoinGroupRequestData request = new JoinGroupRequestBuilder()
+            .withGroupId("group-id")
+            .withMemberId(UNKNOWN_MEMBER_ID)
+            .withDefaultProtocolTypeAndProtocols()
+            .build();
+
+        // First round of join requests. Generate member ids. All requests will be accepted
+        // as the group is still Empty.
+        List<CompletableFuture<JoinGroupResponseData>> firstRoundFutures = new ArrayList<>();
+        IntStream.range(0, groupMaxSize + 1).forEach(i -> {
+            CompletableFuture<JoinGroupResponseData> responseFuture = new CompletableFuture<>();
+            firstRoundFutures.add(responseFuture);
+            CoordinatorResult<Void, Record> result = context.sendGenericGroupJoin(request, responseFuture, requiredKnownMemberId);
+            assertTrue(responseFuture.isDone());
+            try {
+                assertEquals(Errors.MEMBER_ID_REQUIRED.code(), responseFuture.get().errorCode());
+            } catch (Exception ignored) {
```

Comment:
This catch is a bit suspicious here. I suppose that it would also catch the error thrown by `assertEquals`, no?

Reply:
Do we really need to keep the try..catch?

Reply:
this is required if we want to use the streams api. let me know if we should just use the for each loop

Reply:
gotcha. it is fine like this.